### DISCUSSION TITLE
Defining conversions for units

### DISCUSSION THREAD

**Discussion by DillonJ** [spine-tools/SpineOpt.jl#617] [Ideas]
## Objective
Here I try to summarise what we have now, the issues and some ideas for improving it. This is not a proposal - just some ideas which I hope are useful for stimulating some debate:

## Current System

![image](https://user-images.githubusercontent.com/7080191/232090841-193614fc-513a-4708-aa82-6915204c8c42.png)

![image](https://user-images.githubusercontent.com/7080191/232091073-4b353fa4-38e9-42eb-b853-09c0a49720f7.png)

### Key capability requirements:
 - Easily define arbitrary numbers of input and output flows
 - Easily create piecewise affine linear relationships between any two flows
 - Anything more complicated can be done via user_constraints

## Defining Efficiencies

![image](https://user-images.githubusercontent.com/7080191/232091304-a56cf02c-973d-476b-a619-96b88729ea30.png)

 - The `unit__node__node` relationship allows you to constrain two nodes to each other via a number of different parameters
   - **unit_incremental_heat_rate**: Input_flow = `unit_incremental_heat_rate` * output_flow + `unit_idle_heat_rate` * `units_on`.  It can be piecewise linear, used in conjunction with `operating_points` with monotonically increasing coefficients (not enforced). Used in conjunction with `unit_idle_heat_rate` triggers a fixed flow when the unit is online and `unit_start_flow` triggers a flow on a unit start (start fuel consumption)
   - `fix_ratio_out_in_unit_flow`: equivalent to an efficiency. Output_flow = `fix_ratio_out_in_unit_flow` x output_flow + `fix_units_on_coefficient_out_in` * `units_on`. Ordering of the nodes in the `unit__node__node` relationship matters. The first node will be the output flow and the second node will be treated as the input flow (consistently with the out_in in the parameter name. A units_on coefficient is added with `fix_units_on_coefficient_out_in`, 
   - In addition to `fix_ratio_out_in_unit_flow` you have [constraint]_ratio_[direction1]_[direction2]_unit_flow where constraint can be min, max or fix and determines the sense of the constraint (max: <, min: >, fix: =) while direction1 and direction2 are used to interpret the direction of the flows involved. `In` signifies an input flow to the unit while `out` signifies an output flow from the unit. For each of these parameters, there is a corresponding [constraint]_[direction1]_[direction2]_units_on_coefficient. For example: `max_ratio_in_out_unit_flow` creates the following constraint:
`input_flow` < `max_ratio_in_out_unit_flow` * output_flow + `max_units_on_coefficient_in_out` * `units_on`

## Real World Example: Compressed Air Energy Storage
To give a feeling for why these functionalities are useful, we actually use most of them in our models and here is one example:

![image](https://user-images.githubusercontent.com/7080191/232092326-f37f69f4-7f61-470d-a557-ff5804c3a13e.png)

## Issues with the Current System

 - Multiple ways to do the same thing (kind of)
 - The ordering of nodes in `unit__node__node` relationship matters and this can be confusing
 - When specifying a `unit__node__node` relationship, currently toolbox doesn’t constrain a user to choosing nodes that are connected to the unit. It’s possible to create a `unit__node__node` relationship between a unit and nodes where there are no flows. We actually need to define a relationship between two flows, which is really a relationship between a unit__[to/from]_node relationship and a unit__[to/from]_node relationship.
 - There is a long list of parameters (24 in total) ([fix/max/min]_ratio_[in/out]_[in/out]_[unit_flow/units_on_coefficient]
 - `Incremental_heat_rate` supports piecewise linear but the ratio constraints don’t

## Idea for the simple case:

![image](https://user-images.githubusercontent.com/7080191/232093794-8545b1f7-2693-456e-bb6d-200441b6d783.png)

 - Compliments the existing structure for the simplest case.
 - Works with multiple inputs (fuel switching or retrofitting)
 - Could implement a piecewise-linear formulation

However,
 - Doesn’t allow the full flexibility to create relationships between any two flows.
 - Would have to be used in conjunction with another option and that might be confusing?
 - This could work for a user if they have only a simple input and output - but adding a third flow might mess things up without them realising

## Idea: Specifying Conversions Relative to a Reference Flow

![image](https://user-images.githubusercontent.com/7080191/232093062-0f997ebc-79d8-4051-a74e-694f94fd0828.png)

## Idea: Specifying Conversions Relative to a Reference Commodity

![image](https://user-images.githubusercontent.com/7080191/232093168-84eea688-bc1e-4044-a68d-cdba47d68a39.png)

**Comment by manuelma** [comment]
> We actually need to define a relationship between two flows, which is really a relationship between a unit__[to/from]node relationship and a unit_[to/from]_node relationship.

But not *any* two flows. It needs to be two flows for the same unit. This can't be solved directly by supporting multiD entities between multiD entities - we need something extra, some sort of SQL JOIN implementation in our EAV system.

**Comment by DillonJ** [comment]
Well it doesn’t really matter here if they are the same unit or not, I see that as an additional bonus. It allows us to constrain any two flows from a particular unit which is is basic requirement but is not constrained to flows from a single unit which could be an advantage when implementing composite units, fuel switching and that kind of thing.

However I get that from the user perspective and the in the simple case, it could be desirable to relate the efficiency to a unit
